hydra-cluster/bench/Main.hs

```diff
@@ -114,19 +142,83 @@ main =
           pure $ fromLedgerPParams ShelleyBasedEraShelley (sgProtocolParams shelleyGenesis)
     dataset <- generateConstantUTxODataset pparams (fromIntegral clusterSize) numberOfTxs
     saveDataset benchDir dataset
-    run timeoutSeconds benchDir dataset clusterSize
+    run options benchDir dataset
 
-  replay Options{timeoutSeconds, clusterSize} benchDir = do
+  replay options benchDir = do
     datasets <- either die pure =<< eitherDecodeFileStrict' (benchDir </> "dataset.json")
     putStrLn $ "Using UTxO and Transactions from: " <> benchDir
-    run timeoutSeconds benchDir datasets clusterSize
+    run options benchDir datasets
 
-  -- TODO(SN): Ideally we would like to say "to re-run use ... " on errors
-  run timeoutSeconds benchDir datasets clusterSize = do
+  run options@Options{timeoutSeconds, clusterSize, startingNodeId} benchDir datasets = do
     putStrLn $ "Test logs available in: " <> (benchDir </> "test.log")
-    withArgs [] . hspec $ bench timeoutSeconds benchDir datasets clusterSize
+    withArgs [] $
+      try (bench startingNodeId timeoutSeconds benchDir datasets clusterSize) >>= \case
+        Left (err :: HUnitFailure) ->
+          benchmarkFailedWith benchDir err
+        Right summary ->
+          benchmarkSucceeded options benchDir summary
 
   saveDataset tmpDir dataset = do
     let txsFile = tmpDir </> "dataset.json"
     putStrLn $ "Writing dataset to: " <> txsFile
     encodeFile txsFile dataset
+
+benchmarkFailedWith :: FilePath -> HUnitFailure -> IO ()
+benchmarkFailedWith benchDir (HUnitFailure _ reason) = do
```

Comment:
Should: print the source location as we are not using `hspec` anymore.

Why? It was printed before and was at least an indication which assertion failed by the collected `HasCallStack`.

Alternative: use `[hspecResult`](https://hackage.haskell.org/package/hspec-core-2.11.1/docs/Test-Hspec-Core-Runner.html#v:hspecResult) as this should (AFAIS, but not tested) print the output as before and you can act on the outcome.

Reply:
`bench` is not a `Spec` anymore so using hspecResult won't work I think.